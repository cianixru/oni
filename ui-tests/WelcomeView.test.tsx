import { shallow } from "enzyme"
import toJson from "enzyme-to-json"
import { Event } from "oni-types"
import * as React from "react"

import {
    WelcomeView,
    WelcomeViewProps,
    WelcomeViewState,
    IWelcomeInputEvent,
    SessionsList,
    SectionItem,
    SectionHeader,
} from "./../browser/src/Editor/NeovimEditor/WelcomeBufferLayer"

describe("<WelcomeView />", () => {
    const buttons = [
        "button1",
        "button2",
        "button3",
        "button4",
        "button5",
        "button6",
        "button7",
        "button8",
    ]

    const restoreSession = jest.fn()
    const executeCommand = jest.fn()

    const commands = {
        openFile: {
            execute: jest.fn(),
            command: "button1",
        },
        openTutor: {
            execute: jest.fn(),
            command: "button2",
        },
        openDocs: {
            execute: jest.fn(),
            command: "button3",
        },
        openConfig: {
            execute: jest.fn(),
            command: "button4",
        },
        openThemes: {
            execute: jest.fn(),
            command: "button5",
        },
        openWorkspaceFolder: {
            execute: jest.fn(),
            command: "button6",
        },
        quickOpenShow: {
            execute: jest.fn(),
            command: "button7",
        },
        commandline: {
            execute: jest.fn(),
            command: "button8",
        },
        restoreSession,
    }

    const sessions = [
        {
            name: "test",
            id: "test-1",
            file: "/test/dir",
            directory: "/test/dir",
            workspace: "/test/dir",
        },
    ]

    const inputEvent = new Event<IWelcomeInputEvent>("handleInputTestEvent")
    const getMetadata = async () => ({ version: "1" })

    const ids = [...buttons, ...sessions.map(({ id }) => id)]
    const wrapper = shallow<WelcomeViewProps, WelcomeViewState>(
        <WelcomeView
            active
            ids={ids}
            sessions={sessions}
            commands={commands}
            inputEvent={inputEvent}
            getMetadata={getMetadata}
            restoreSession={restoreSession}
            sections={[buttons.length, sessions.length]}
            executeCommand={executeCommand}
        />,
    )

    it("Should render without crashing", () => {
        expect(wrapper.length).toBe(1)
    })

    it("should match its last snapshot", () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it("should set the correct version", () => {
        expect(wrapper.state().version).toBe("1")
    })

    it("should have the correct initial state", () => {
        expect(wrapper.state()).toEqual({ version: "1", currentIndex: 0, selectedId: "button1" })
    })

    it("should correctly update selection based on input", () => {
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 1, select: false })
        expect(wrapper.state().selectedId).toBe("button2")
    })

    it("should loop back to last button if user navigates upwards at the first button", () => {
        wrapper.setState({ currentIndex: 0, selectedId: "button1" })
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: -1, select: false })
        expect(wrapper.state().currentIndex).toBe(ids.length - 1)
        expect(wrapper.state().selectedId).toBe("test-1")
    })

    it("should loop back to first button if user navigates downwards at the last button", () => {
        wrapper.setState({ currentIndex: ids.length - 1, selectedId: "test-1" })
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 1, select: false })
        expect(wrapper.state().currentIndex).toBe(0)
        expect(wrapper.state().selectedId).toBe("button1")
    })

    it("should trigger a command on enter event", () => {
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 0, select: true })
        expect(commands.openFile.execute).toHaveBeenCalled()
    })

    it("should navigate right if horizontal is 1", () => {
        wrapper.setState({ currentIndex: 7, selectedId: "button8" })
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 0, horizontal: 1, select: false })
        expect(wrapper.state().currentIndex).toBe(8)
        expect(wrapper.state().selectedId).toBe("test-1")
    })

    it("should navigate left if horizontal is -1", () => {
        wrapper.setState({ currentIndex: ids.length - 1, selectedId: "test-1" })
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 0, horizontal: -1, select: false })
        expect(wrapper.state().currentIndex).toBe(0)
        expect(wrapper.state().selectedId).toBe("button1")
    })

    it("should call restore session if the selectedId is a session", () => {
        wrapper.setState({ currentIndex: ids.length - 1, selectedId: "test-1" })
        const instance = wrapper.instance() as WelcomeView
        instance.handleInput({ vertical: 0, horizontal: 0, select: true })
        expect(restoreSession.mock.calls[0][0]).toBe("test")
    })

    it("Should render a section header", () => {
        const headers = wrapper.dive().find(SectionHeader).length
        expect(headers).toBe(1)
    })

    it("should render the place holder text if there are no sessions available", () => {
        wrapper.setProps({ sessions: [] })
        const placeholder = wrapper
            .find(SectionItem)
            .dive()
            .text()
        expect(placeholder).toBe("No Sessions Available")
    })
})
