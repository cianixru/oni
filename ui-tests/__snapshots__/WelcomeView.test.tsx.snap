// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WelcomeView /> should match its last snapshot 1`] = `
<styled.div
  data-id="welcome-screen"
  height="100%"
  innerRef={
    Object {
      "current": null,
    }
  }
>
  <styled.div
    extension={
      Array [
        "
    width: 100%;
    padding-top: 32px;
    animation: ",
        "ckrRMN",
        " 0.25s ease-in 0.25s forwards};
",
      ]
    }
  >
    <styled.div />
    <styled.div
      alignment="flex-end"
    >
      <styled.div>
        Oni
      </styled.div>
      <styled.div>
        Modern Modal Editing
      </styled.div>
    </styled.div>
    <styled.div
      flex="0 0"
    >
      <styled.img
        src="images/oni-icon-no-border.svg"
      />
    </styled.div>
    <styled.div
      alignment="flex-start"
    >
      <div>
        https://onivim.io
      </div>
    </styled.div>
    <styled.div />
  </styled.div>
  <styled.div
    extension={
      Array [
        "
    width: 60%;
    height: 60%;
    padding: 0 1em;
    opacity: 1;
    margin-top: 64px;
    box-sizing: border-box;
    border: 1px solid ",
        [Function],
        ";
    border-radius: 4px;
    overflow: hidden;
    justify-content: space-around;
    background-color: ",
        [Function],
        ";
    ",
        "
    box-shadow: inset 0 4px 8px 2px rgba(0, 0, 0, 0.2);
",
        ";
",
      ]
    }
  >
    <Component
      commands={
        Object {
          "commandline": Object {
            "command": "button8",
            "execute": [MockFunction],
          },
          "openConfig": Object {
            "command": "button4",
            "execute": [MockFunction],
          },
          "openDocs": Object {
            "command": "button3",
            "execute": [MockFunction],
          },
          "openFile": Object {
            "command": "button1",
            "execute": [MockFunction],
          },
          "openThemes": Object {
            "command": "button5",
            "execute": [MockFunction],
          },
          "openTutor": Object {
            "command": "button2",
            "execute": [MockFunction],
          },
          "openWorkspaceFolder": Object {
            "command": "button6",
            "execute": [MockFunction],
          },
          "quickOpenShow": Object {
            "command": "button7",
            "execute": [MockFunction],
          },
          "restoreSession": [MockFunction],
        }
      }
      executeCommand={[MockFunction]}
      selectedId="button1"
    />
    <styled.div>
      <styled.ul>
        <styled.div>
          Sessions
        </styled.div>
        <styled.li
          isSelected={false}
          key="test-1"
          onClick={[Function]}
        >
          <Icon
            name="file"
            style={
              Object {
                "marginRight": "0.3em",
              }
            }
          />
           
          test
        </styled.li>
      </styled.ul>
    </styled.div>
  </styled.div>
</styled.div>
`;
